README.md
# PIA installer for Fedora (UNOFFICIAL)

*Disclaimer: Installation and use of any software made by third party developers is at your own discretion and
liability. I'm sharing all changes made to the official installer and doing my best to make sure it's safe
to use for fellow fedora users.*

**Pull requests and issues are welcomed here**

[Changelog](CHANGELOG.md).

## Other distros
"drhedberg" reported that the script also works with Opensuse 13.2 by replacing dnf with zypper.
"edouardmenayde" [solution](https://github.com/edouardmenayde/pia-fedora) for v81 support.

## Updated for v81
Based off of the official installer 
[here](https://www.privateinternetaccess.com/installer/download_installer_linux)
I had to do some changes to the original installer to install required packages 
via dnf. Check details below for detailed changes.

### Versions tested

* Fedora 27 Workstation.

Thanks @edouardmenayde for the [solution](https://github.com/edouardmenayde/pia-fedora). 
I was able to put these changes into the installer
to simplify installation process. You can also use this same installer for all ready supported
systems(ubuntu, arch, etc...).

### Instructions
Download installer:

[https://storage.googleapis.com/installers-pia-unofficial/pia-v81-installer-fedora.tar.gz]

Make sure MD5 is: d153b88e20256c53ff4c1ecc4110e535

```
$md5sum pia-v81-installer-fedora.tar.gz 
d153b88e20256c53ff4c1ecc4110e535  pia-v81-installer-fedora.tar.gz
```


then follow official instalation guide 
[here](https://www.privateinternetaccess.com/installer/download_installer_linux)

or
```
tar -xzvf pia-v81-installer-fedora.tar.gz
./pia-v81-installer-fedora.sh
```
then it will prompt for password since it'll install necessary packages via dnf

### Specific Changes
For those of you who care:

In line 155 installer.rb add the following lines:
```
155       elsif system("sudo which dnf > /dev/null 2>&1")
156         command = "sudo dnf install -y GConf2 net-tools libappindicator libXScrnSaver"
157         progress "Running: #{command}"
158         if ! system(command)
159           progress "Package installation failed. Please make sure that your package manager is configured correctly and is still supporte
160           exit 1
161         end
```
These line install dependencies for fedora systems. Official installer supports Pacman(Archlinux). I'm not
sure why fedora is not supported...

## For PIA V61

Automatically setup Private Internet Access VPN v65 for Fedora users. This is an UNOFFICIAL installer that originally was based on the OFFICIAL Ubuntu install script. Feel free to check the src.

## Versions supported.

* Fedora 22 desktop
* Fedora 23 Workstation.
* Fedora 24 Workstation.
* Fedora 25 Workstation.
* Fedora 26 Workstation.
* Fedora 27 Workstation.

It may work with other versions. This script has not been tested with earlier edition Fedora.

_**NOTE**: this script uses dnf._


## Install

```
git clone https://github.com/shaynewang/piafedora.git
cd piafedora
chmod +x install_fedora.sh
sudo ./install_fedora.sh
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.2]
### Changed
- Updated for PIA v81
- Added download link for v81 installer(unofficial)

## [1.1]
### Added
- Upgrade support for old versions.
- This file :)

### Changed
- Added new PIA CA cert.
- Update PIA v.65

## [1.0]
- Original version which is a port of the Ubuntu script
- Added Universal password


install_redhat7.sh
#! /bin/sh -
#
# Install OpenVPN connections for all available
# regions to NetworkManager
#
# Requirements:
#   should be run as root
#   python and openvpn (will be installed if not present)
#
# Usage:
#  install [--version]

IFS='
	 '
SERVER_INFO=/tmp/server_info
SPLIT_TOKEN=':'

error( )
{
  echo "$@" 1>&2
  exit 1
}

error_and_usage( )
{
  echo "$@" 1>&2
  usage_and_exit 1
}

usage( )
{
  echo "Usage: sudo `dirname $0`/$PROGRAM"
}

usage_and_exit( )
{
  usage
  exit $1
}

version( )
{
  echo "$PROGRAM version $VERSION"
}

read_user_login( )
{
  echo -n "Please enter your login: "
  read LOGIN
  if [ -z $LOGIN ]; then
    error "A login must be provided for the installation to proceed"
  fi

  echo "Do you want to include your password?"

  select yn in "Yes" "No"; do
	  case $yn in
		  "Yes" )	echo -n "Please enter your password: "
				read PASS
				break;;
		  "No" )	exit;;
	  esac
  done

  echo $PASS
}

verify_running_as_root( )
{
  if [ `/usr/bin/id -u` -ne 0 ]; then
      error_and_usage "$0 must be run as root"
  fi
}

install_python_version( )
{
  if ! python -V; then
    echo -n 'Package python2.7 required. Install? (y/n): '
    read install_python
    if [ $install_python = 'y' ]; then
      echo "Installing python2.7.."
      if ! yum install python; then
        error "Error installing python2.7 Aborting.."
      fi
    else
      error "Package python2.7 is required for installation. Aborting.."
    fi
  else
    echo "Package python2.7 already installed"
  fi
}

install_open_vpn( )
{
  if ! yum search NetworkManager-openvpn-gnome; then
    echo -n 'Package NetworkManager-openvpn-gnome required. Install? (y/n): '
    read install_openvpn
    if [ $install_openvpn = 'y' ]; then
      echo "Installing network-manager-openvpn.."
      if ! yum install NetworkManager-openvpn NetworkManager-openvpn-gnome -y; then
        error "Error installing network-manager-openvpn. Aborting.."
      fi
    else
      error "Package network-manager-openvpn is required for installation. Aborting.."
    fi
  else
    echo "Package network-manager-openvpn already installed"
  fi
}

copy_crt( )
{
  echo 'Copying certificate..'
  mkdir -p /etc/openvpn
cat << EOF > /etc/openvpn/ca.crt
-----BEGIN CERTIFICATE-----
MIID2jCCA0OgAwIBAgIJAOtqMkR2JSXrMA0GCSqGSIb3DQEBBQUAMIGlMQswCQYD
VQQGEwJVUzELMAkGA1UECBMCT0gxETAPBgNVBAcTCENvbHVtYnVzMSAwHgYDVQQK
ExdQcml2YXRlIEludGVybmV0IEFjY2VzczEjMCEGA1UEAxMaUHJpdmF0ZSBJbnRl
cm5ldCBBY2Nlc3MgQ0ExLzAtBgkqhkiG9w0BCQEWIHNlY3VyZUBwcml2YXRlaW50
ZXJuZXRhY2Nlc3MuY29tMB4XDTEwMDgyMTE4MjU1NFoXDTIwMDgxODE4MjU1NFow
gaUxCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJPSDERMA8GA1UEBxMIQ29sdW1idXMx
IDAeBgNVBAoTF1ByaXZhdGUgSW50ZXJuZXQgQWNjZXNzMSMwIQYDVQQDExpQcml2
YXRlIEludGVybmV0IEFjY2VzcyBDQTEvMC0GCSqGSIb3DQEJARYgc2VjdXJlQHBy
aXZhdGVpbnRlcm5ldGFjY2Vzcy5jb20wgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJ
AoGBAOlVlkHcxfN5HAswpryG7AN9CvcvVzcXvSEo91qAl/IE8H0knKZkIAhe/z3m
hz0t91dBHh5yfqwrXlGiyilplVB9tfZohvcikGF3G6FFC9j40GKP0/d22JfR2vJt
4/5JKRBlQc9wllswHZGmPVidQbU0YgoZl00bAySvkX/u1005AgMBAAGjggEOMIIB
CjAdBgNVHQ4EFgQUl8qwY2t+GN0pa/wfq+YODsxgVQkwgdoGA1UdIwSB0jCBz4AU
l8qwY2t+GN0pa/wfq+YODsxgVQmhgaukgagwgaUxCzAJBgNVBAYTAlVTMQswCQYD
VQQIEwJPSDERMA8GA1UEBxMIQ29sdW1idXMxIDAeBgNVBAoTF1ByaXZhdGUgSW50
ZXJuZXQgQWNjZXNzMSMwIQYDVQQDExpQcml2YXRlIEludGVybmV0IEFjY2VzcyBD
QTEvMC0GCSqGSIb3DQEJARYgc2VjdXJlQHByaXZhdGVpbnRlcm5ldGFjY2Vzcy5j
b22CCQDrajJEdiUl6zAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAByH
atXgZzjFO6qctQWwV31P4qLelZzYndoZ7olY8ANPxl7jlP3YmbE1RzSnWtID9Gge
fsKHi1jAS9tNP2E+DCZiWcM/5Y7/XKS/6KvrPQT90nM5klK9LfNvS+kFabMmMBe2
llQlzAzFiIfabACTQn84QLeLOActKhK8hFJy2Gy6
-----END CERTIFICATE-----
EOF

}

fix_conflict( )
{
  semanage fcontext -a -t home_cert_t /etc/openvpn/ca.crt
  restorecon -R -v /etc/openvpn/ca.crt
}

parse_server_info( )
{
  echo 'Loading servers information..'
  json=`wget -q -O - 'https://privateinternetaccess.com/vpninfo/servers?version=24' | head -1`

  python > $SERVER_INFO <<EOF
payload = '$json'
import json
d = json.loads(payload)
print "\n".join([d[k]['name']+'$SPLIT_TOKEN'+d[k]['dns'] for k in d.keys() if k != 'info'])
EOF
}

write_config_files( )
{
  echo 'Removing previous config files if existing..'
  rm -f /etc/NetworkManager/system-connections/PIA\ -\ *

  echo 'Creating config files..'
  IFS='
'
  while read server_info; do
    name="PIA - `echo $server_info | awk -F: '{print $1}'`"
    dns=`echo $server_info | awk -F: '{print $2}'`
    cat <<EOF > /etc/NetworkManager/system-connections/$name
[connection]
id=$name
uuid=`uuidgen`
type=vpn
autoconnect=false

[vpn]
service-type=org.freedesktop.NetworkManager.openvpn
username=$LOGIN
comp-lzo=yes
remote=$dns
connection-type=password
password-flags=0
ca=/etc/openvpn/ca.crt

[vpn-secrets]
password=$PASS

[ipv4]
method=auto
EOF
  chmod 600 /etc/NetworkManager/system-connections/$name
  done < $SERVER_INFO
  rm $SERVER_INFO
  IFS='
 	'
}

restart_network_manager( )
{
  echo 'Restarting network manager..'
  systemctl restart NetworkManager
}

EXITCODE=0
PROGRAM=`basename $0`
VERSION=1.0

while test $# -gt 0
do
  case $1 in
  --usage | --help | -h )
    usage_and_exit 0
    ;;
  --version | -v )
    version
    exit 0
    ;;
  *)
    error_and_usage "Unrecognized option: $1"
    ;;
  esac
  shift
done


verify_running_as_root
install_python_version
install_open_vpn
read_user_login
copy_crt
fix_conflict
parse_server_info
write_config_files
restart_network_manager

echo "Install successful!"
exit 0



install_fedora22.sh
#! /bin/sh -
#
# Install OpenVPN connections for all available
# regions to NetworkManager
#
# Requirements:
#   should be run as root
#   python and openvpn (will be installed if not present)
#
# Usage:
#  install [--version]

IFS='
	 '
SERVER_INFO=/tmp/server_info
SPLIT_TOKEN=':'

error( )
{
  echo "$@" 1>&2
  exit 1
}

error_and_usage( )
{
  echo "$@" 1>&2
  usage_and_exit 1
}

usage( )
{
  echo "Usage: sudo `dirname $0`/$PROGRAM"
}

usage_and_exit( )
{
  usage
  exit $1
}

version( )
{
  echo "$PROGRAM version $VERSION"
}

read_user_login( )
{
  echo -n "Please enter your login: "
  read LOGIN
  if [ -z $LOGIN ]; then
    error "A login must be provided for the installation to proceed"
  fi

  echo "Do you want to include your password?"

  select yn in "Yes" "No"; do
	  case $yn in
		  "Yes" )	echo -n "Please enter your password: "
				read PASS
				break;;
		  "No" )	exit;;
	  esac
  done

  echo $PASS
}

verify_running_as_root( )
{
  if [ `/usr/bin/id -u` -ne 0 ]; then
      error_and_usage "$0 must be run as root"
  fi
}

install_python_version( )
{
  if ! python -V; then
    echo -n 'Package python2.7 required. Install? (y/n): '
    read install_python
    if [ $install_python = 'y' ]; then
      echo "Installing python2.7.."
      if ! dnf install python; then
        error "Error installing python2.7 Aborting.."
      fi
    else
      error "Package python2.7 is required for installation. Aborting.."
    fi
  else
    echo "Package python2.7 already installed"
  fi
}

install_open_vpn( )
{
  if ! dnf search NetworkManager-openvpn-gnome; then
    echo -n 'Package NetworkManager-openvpn-gnome required. Install? (y/n): '
    read install_openvpn
    if [ $install_openvpn = 'y' ]; then
      echo "Installing network-manager-openvpn.."
      if ! dnf install NetworkManager-openvpn-gnome; then
        error "Error installing network-manager-openvpn. Aborting.."
      fi
    else
      error "Package network-manager-openvpn is required for installation. Aborting.."
    fi
  else
    echo "Package network-manager-openvpn already installed"
  fi
}

copy_crt( )
{
  echo 'Copying certificate..'
  mkdir -p /etc/openvpn
cat << EOF > /etc/openvpn/ca.crt
-----BEGIN CERTIFICATE-----
MIID2jCCA0OgAwIBAgIJAOtqMkR2JSXrMA0GCSqGSIb3DQEBBQUAMIGlMQswCQYD
VQQGEwJVUzELMAkGA1UECBMCT0gxETAPBgNVBAcTCENvbHVtYnVzMSAwHgYDVQQK
ExdQcml2YXRlIEludGVybmV0IEFjY2VzczEjMCEGA1UEAxMaUHJpdmF0ZSBJbnRl
cm5ldCBBY2Nlc3MgQ0ExLzAtBgkqhkiG9w0BCQEWIHNlY3VyZUBwcml2YXRlaW50
ZXJuZXRhY2Nlc3MuY29tMB4XDTEwMDgyMTE4MjU1NFoXDTIwMDgxODE4MjU1NFow
gaUxCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJPSDERMA8GA1UEBxMIQ29sdW1idXMx
IDAeBgNVBAoTF1ByaXZhdGUgSW50ZXJuZXQgQWNjZXNzMSMwIQYDVQQDExpQcml2
YXRlIEludGVybmV0IEFjY2VzcyBDQTEvMC0GCSqGSIb3DQEJARYgc2VjdXJlQHBy
aXZhdGVpbnRlcm5ldGFjY2Vzcy5jb20wgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJ
AoGBAOlVlkHcxfN5HAswpryG7AN9CvcvVzcXvSEo91qAl/IE8H0knKZkIAhe/z3m
hz0t91dBHh5yfqwrXlGiyilplVB9tfZohvcikGF3G6FFC9j40GKP0/d22JfR2vJt
4/5JKRBlQc9wllswHZGmPVidQbU0YgoZl00bAySvkX/u1005AgMBAAGjggEOMIIB
CjAdBgNVHQ4EFgQUl8qwY2t+GN0pa/wfq+YODsxgVQkwgdoGA1UdIwSB0jCBz4AU
l8qwY2t+GN0pa/wfq+YODsxgVQmhgaukgagwgaUxCzAJBgNVBAYTAlVTMQswCQYD
VQQIEwJPSDERMA8GA1UEBxMIQ29sdW1idXMxIDAeBgNVBAoTF1ByaXZhdGUgSW50
ZXJuZXQgQWNjZXNzMSMwIQYDVQQDExpQcml2YXRlIEludGVybmV0IEFjY2VzcyBD
QTEvMC0GCSqGSIb3DQEJARYgc2VjdXJlQHByaXZhdGVpbnRlcm5ldGFjY2Vzcy5j
b22CCQDrajJEdiUl6zAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAByH
atXgZzjFO6qctQWwV31P4qLelZzYndoZ7olY8ANPxl7jlP3YmbE1RzSnWtID9Gge
fsKHi1jAS9tNP2E+DCZiWcM/5Y7/XKS/6KvrPQT90nM5klK9LfNvS+kFabMmMBe2
llQlzAzFiIfabACTQn84QLeLOActKhK8hFJy2Gy6
-----END CERTIFICATE-----
EOF

}

fix_conflict( )
{
  semanage fcontext -a -t home_cert_t /etc/openvpn/ca.crt
  restorecon -R -v /etc/openvpn/ca.crt
}

parse_server_info( )
{
  echo 'Loading servers information..'
  json=`wget -q -O - 'https://privateinternetaccess.com/vpninfo/servers?version=24' | head -1`

  python > $SERVER_INFO <<EOF
payload = '$json'
import json
d = json.loads(payload)
print "\n".join([d[k]['name']+'$SPLIT_TOKEN'+d[k]['dns'] for k in d.keys() if k != 'info'])
EOF
}

write_config_files( )
{
  echo 'Removing previous config files if existing..'
  rm -f /etc/NetworkManager/system-connections/PIA\ -\ *

  echo 'Creating config files..'
  IFS='
'
  while read server_info; do
    name="PIA - `echo $server_info | awk -F: '{print $1}'`"
    dns=`echo $server_info | awk -F: '{print $2}'`
    cat <<EOF > /etc/NetworkManager/system-connections/$name
[connection]
id=$name
uuid=`uuidgen`
type=vpn
autoconnect=false

[vpn]
service-type=org.freedesktop.NetworkManager.openvpn
username=$LOGIN
comp-lzo=yes
remote=$dns
connection-type=password
password-flags=0
ca=/etc/openvpn/ca.crt

[vpn-secrets]
password=$PASS

[ipv4]
method=auto
EOF
  chmod 600 /etc/NetworkManager/system-connections/$name
  done < $SERVER_INFO
  rm $SERVER_INFO
  IFS='
 	'
}

restart_network_manager( )
{
  echo 'Restarting network manager..'
  systemctl restart NetworkManager
}

EXITCODE=0
PROGRAM=`basename $0`
VERSION=1.0

while test $# -gt 0
do
  case $1 in
  --usage | --help | -h )
    usage_and_exit 0
    ;;
  --version | -v )
    version
    exit 0
    ;;
  *)
    error_and_usage "Unrecognized option: $1"
    ;;
  esac
  shift
done


verify_running_as_root
install_python_version
install_open_vpn
read_user_login
copy_crt
fix_conflict
parse_server_info
write_config_files
restart_network_manager

echo "Install successful!"
exit 0



install_fedora.sh
#! /bin/sh -
#
# Install OpenVPN connections for all available
# regions to NetworkManager
#
# Requirements:
#   should be run as root
#   python and openvpn (will be installed if not present)
#
# Usage:
#  install [--version]

IFS='
	 '
SERVER_INFO=/tmp/server_info
SPLIT_TOKEN=':'

error( )
{
  echo "$@" 1>&2
  exit 1
}

error_and_usage( )
{
  echo "$@" 1>&2
  usage_and_exit 1
}

usage( )
{
  echo "Usage: sudo `dirname $0`/$PROGRAM"
}

usage_and_exit( )
{
  usage
  exit $1
}

version( )
{
  echo "$PROGRAM version $VERSION"
}

read_user_login( )
{
  echo -n "Please enter your login: "
  read LOGIN
  if [ -z $LOGIN ]; then
    error "A login must be provided for the installation to proceed"
  fi

  echo "Do you want to include your password?"

  select yn in "Yes" "No"; do
	  case $yn in
		  "Yes" )	echo -n "Please enter your password: "
				read PASS
				break;;
		  "No" )	exit;;
	  esac
  done

  echo $PASS
}

verify_running_as_root( )
{
  if [ `/usr/bin/id -u` -ne 0 ]; then
      error_and_usage "$0 must be run as root"
  fi
}

install_python_version( )
{
  if ! python -V; then
    echo -n 'Package python2.7 required. Install? (y/n): '
    read install_python
    if [ $install_python = 'y' ]; then
      echo "Installing python2.7.."
      if ! dnf install python; then
        error "Error installing python2.7 Aborting.."
      fi
    else
      error "Package python2.7 is required for installation. Aborting.."
    fi
  else
    echo "Package python2.7 already installed"
  fi
}

install_open_vpn( )
{
  if ! dnf search NetworkManager-openvpn-gnome; then
    echo -n 'Package NetworkManager-openvpn-gnome required. Install? (y/n): '
    read install_openvpn
    if [ $install_openvpn = 'y' ]; then
      echo "Installing network-manager-openvpn.."
      if ! dnf install NetworkManager-openvpn-gnome; then
        error "Error installing network-manager-openvpn. Aborting.."
      fi
    else
      error "Package network-manager-openvpn is required for installation. Aborting.."
    fi
  else
    echo "Package network-manager-openvpn already installed"
  fi
}

install_crt( )
{
  echo 'Downloading certificate..'
  wget -O - https://www.privateinternetaccess.com/openvpn/openvpn.zip > /tmp/openvpn.zip
  echo 'Installing certificate..'
  mkdir -p /etc/openvpn
  unzip -p /tmp/openvpn.zip ca.rsa.2048.crt > /etc/openvpn/pia_ca.rsa.2048.crt
}

fix_conflict( )
{
  semanage fcontext -a -t home_cert_t /etc/openvpn/pia_ca.rsa.2048.crt
  restorecon -R -v /etc/openvpn/pia_ca.rsa.2048.crt
}

parse_server_info( )
{
  echo 'Loading servers information..'
  json=`wget -q -O - 'https://privateinternetaccess.com/vpninfo/servers?version=24' | head -1`

  python > $SERVER_INFO <<EOF
payload = '$json'
import json
d = json.loads(payload)
print "\n".join([d[k]['name']+'$SPLIT_TOKEN'+d[k]['dns'] for k in d.keys() if k != 'info'])
EOF
}

write_config_files( )
{
  echo 'Removing previous config files if existing..'
  rm -f /etc/NetworkManager/system-connections/PIA\ -\ *

  echo 'Creating config files..'
  IFS='
'
  while read server_info; do
    name="PIA - `echo $server_info | awk -F: '{print $1}'`"
    dns=`echo $server_info | awk -F: '{print $2}'`
    cat <<EOF > /etc/NetworkManager/system-connections/$name
[connection]
id=$name
uuid=`uuidgen`
type=vpn
autoconnect=false

[vpn]
service-type=org.freedesktop.NetworkManager.openvpn
username=$LOGIN
comp-lzo=yes
remote=$dns
connection-type=password
password-flags=0
ca=/etc/openvpn/pia_ca.rsa.2048.crt
cipher=AES-128-CBC
port=1198

[vpn-secrets]
password=$PASS

[ipv4]
method=auto
EOF
  chmod 600 /etc/NetworkManager/system-connections/$name
  done < $SERVER_INFO
  rm $SERVER_INFO
  IFS='
 	'
}

check_previous_config( )
{
  if ls /etc/NetworkManager/system-connections/PIA* 1> /dev/null 2>&1; then
    echo -n 'PIA already installed. Try to update? (NOTE: this will remove the original configuration) (y/n): '
    read remove_old_config
    if [ $remove_old_config = 'y' ]; then
      echo "Removing old config.."
      # TODO: Backup old config
      rm -f /etc/NetworkManager/system-connections/PIA*
    else
      error "Aborting.."
    fi
  fi
}

restart_network_manager( )
{
  echo 'Restarting network manager..'
  systemctl restart NetworkManager
}

EXITCODE=0
PROGRAM=`basename $0`
VERSION=1.1

while test $# -gt 0
do
  case $1 in
  --usage | --help | -h )
    usage_and_exit 0
    ;;
  --version | -v )
    version
    exit 0
    ;;
  *)
    error_and_usage "Unrecognized option: $1"
    ;;
  esac
  shift
done


verify_running_as_root
install_python_version
install_open_vpn
read_user_login
install_crt
fix_conflict
parse_server_info
check_previous_config
write_config_files
restart_network_manager

echo "Install successful!"
exit 0
